app/helpers/icons.py

```diff
@@ -0,0 +1,38 @@
+import os
+
+image_folder = os.path.abspath(os.path.join(os.path.dirname(__file__), '../../static/images/'))
+
+
+def get_all_icons(base_url, red=255, green=0, blue=0):
+    """
+    List all available icons and return them as dictionary entries with the following structure::
+
+        {
+            "category": "default|any folder found in static/images",
+            "icon": "file name of the icon (with .png extension)",
+            "url": "full URL to access this icon",
+        }
+
+    It will use red as the default color if none is defined as argument
+
+    Args:
+        base_url: Base URL of this server (e.g. http://localhost:5000 when doing a local serve, or
+            any other DNS name this server has). This value can be extracted from the (flask)
+            request itself by using `request.base_url`.
+        red: red value for all icons' URL, between 0 and 255 (default: 255)
+        green: green value for all icons' URL, between 0 and 255 (default: 0)
+        blue: blue value for all icons' URL, between 0 and 255 (default: 0)
+    Returns:
+        A list of all available icons, described as dictionary entries (can be then later easily
+            converted to JSON)
+    """
+    icon_sets = []
+    for root, files in os.walk(image_folder):
+        for name in files:
+            category = root.split(os.path.sep)[-1]
+            icon_sets.append({
+                "category": category,
+                "icon": name,
+                "url": "{0}/{1}/{2},{3},{4}/{5}".format(base_url, category, red, green, blue, name)
+            })
```

Comment:
List comprehension would be a bit more efficient especially for list with more than 100 of items:

```python
def category(_root):
  return _root.split(os.path.sep)[-1]

icon_sets = [{
  "category": category(root),
  "icon": name,
  "url": f"{request.host}{route_prefix}/{category}/{red},{green},{blue}/{name}"
} for root, dirs, files in os.walk(image_folder) for name in files]
```

But a bit harder to read for python newbee, it may be worth to quickly test the performance on this code snippets  with `timeit` to see if it worth using comprehension list.

Reply:
I don't know how to test such things...
But in the code you propose we will do two times the computation of `category(root)` so maybe the gain is lost there?
What would make it be faster against a standard `for` loop?

Reply:
To test see https://docs.python.org/3/library/timeit.html

Reply:
I've done quickly the test with the code snippet below, and you are right the string manipulation for the category remove the the time that we gain with comprehension. So the for loop is better :-)

```python
import os
import timeit

image_folder = '/home/ltshb/repositories/service-icons/static/images/default'
request_host = 'www.example.com'
route_prefix = '/v4'
red = 255
blue = 0
green = 0

def category(_root):
    return _root.split(os.path.sep)[-1]

def comprehension():
    icon_sets = [{"category": category(root), "icon": name, "url": f"{request_host}{route_prefix}/{category(root)}/{red},{green},{blue}/{name}"} for root, dirs, files in os.walk(image_folder) for name in files]
    
def for_loop():
    icon_sets = []
    for root, dirs, files in os.walk(image_folder):
        for name in files:
            category = root.split(os.path.sep)[-1]
            icon_sets.append({
                "category": category,
                "icon": name,
                "url": f"{request_host}{route_prefix}/{category}/{red},{green},{blue}/{name}"
            })


print(timeit.timeit('comprehension()', number=1000, globals=globals()))
print(timeit.timeit('for_loop()', number=1000, globals=globals()))

# result
0.8106462870018731
0.6521039840008598
```